src/osp/Active/CommonPhysics.cpp

```diff
@@ -0,0 +1,148 @@
+        // Default radius: 1
+        return (4.0f / 3.0f) * pi * scale.x() * scale.x() * scale.x();
+    case ECollisionShape::BOX:
+        // Default width: 2
+        return 2.0f*scale.x() * 2.0f*scale.y() * 2.0f*scale.z();
+    case ECollisionShape::CYLINDER:
+        // Default radius: 1, default height: 2
+        return pi * scale.x() * scale.x() * 2.0f*scale.z();
+    case ECollisionShape::CAPSULE:
+    case ECollisionShape::CONVEX_HULL:
+    case ECollisionShape::TERRAIN:
+    case ECollisionShape::COMBINED:
+    default:
+        std::cout << "Error: unsupported shape for volume calc\n";
+        return 0.0f;
+    }
+}
+
+Matrix3 transform_inertia_tensor(Matrix3 I, float mass, Vector3 translation, Matrix3 rotation)
+{
+    // Apply rotation via similarity transformation
+    I = rotation.transposed() * I * rotation;
+
+    // Translate via tensor generalized parallel axis theorem
+    using Magnum::Math::dot;
+    Vector3 r = translation;
+    Matrix3 outerProductR = {r * r.x(), r * r.y(), r * r.z()};
+    Matrix3 E3 = Matrix3{};
+
+    return I + mass * (dot(r, r) * E3 - outerProductR);
+}
+
+Vector3 collider_inertia_tensor(ECollisionShape shape, Vector3 scale, float mass)
+{
+    switch (shape)
+    {
+    case ECollisionShape::CYLINDER:
+    {
+        // Default cylinder dimensions: radius 1, height 2
+        float height = 2.0f * scale.z();
+        // If sclY != sclX I will be mad
+        float radius = scale.x();
+        return cylinder_inertia_tensor(radius, height, mass);
+    }
+    case ECollisionShape::BOX:
+    {
+        // Default box dimensions: 2x2x2
+        Vector3 dimensions = 2.0f * scale;
+        return cuboid_inertia_tensor(dimensions, mass);
+    }
+    case ECollisionShape::SPHERE:
+    {
+        // Default sphere: radius = 1
+        Vector3 semiaxes = scale;
+        return ellipsoid_inertia_tensor(semiaxes, mass);
+    }
+    case ECollisionShape::CAPSULE:
+    case ECollisionShape::CONVEX_HULL:
+    case ECollisionShape::TERRAIN:
+    case ECollisionShape::COMBINED:
+    default:
+        std::cout << "ERROR: unknown collision shape\n";
+        return Vector3{0.0f};
+    }
+}
+
+constexpr Vector3 cylinder_inertia_tensor(float radius, float height, float mass)
+{
+    float r2 = radius * radius;
+    float h2 = height * height;
+    float xx = (1.0f / 12.0f) * (3.0f * r2 + h2);
+    float yy = xx;
+    float zz = r2 / 2.0f;
+
+    return Vector3{mass * xx, mass * yy, mass * zz};
+}
+
+constexpr Vector3 cuboid_inertia_tensor(const Vector3 dims, float mass)
+{
+    float x2 = dims.x() * dims.x();
+    float y2 = dims.y() * dims.y();
+    float z2 = dims.z() * dims.z();
+
+    float xx = y2 + z2;
+    float yy = x2 + z2;
+    float zz = x2 + y2;
+
+    // Pre-multiply constant to avoid non-constexpr operator*(float, Vector3)
+    float c = (1.0f / 12.0f) * mass;
+    return Vector3{c * xx, c * yy, c * zz};
+}
+
+constexpr Vector3 ellipsoid_inertia_tensor(const Vector3 semiaxes, float mass)
```

Comment:
Constexpr functions in CPP file?

Reply:
Hmm I may have committed something I didn't mean to, I'll check it out tomorrow